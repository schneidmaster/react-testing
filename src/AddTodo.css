.add-todo-btn {
  background-color: #f1f1f1;
  padding: 10px;
  border-radius: 3px;
  font-family: Helvetica, Arial, sans-serif;
  font-size: 14px;
  margin-right: 15px;
}

.add-todo-btn-primary {
  background-color: blue;
  color: white;
}

.add-todo-btn:hover {
  cursor: pointer;
  opacity: 0.8;
}

.add-todo-form {
  border: solid 1px #999;
  border-radius: 3px;
  padding: 15px;
}
