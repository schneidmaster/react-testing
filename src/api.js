const API_HOST = "http://localhost:3001";

export const getJSON = (path) => {
  return fetch(`${API_HOST}${path}`).then((response) => response.json());
};

export const postJSON = (path, payload) => {
  return fetch(`${API_HOST}${path}`, {
    method: "POST",
    headers: {
      "content-type": "application/json"
    },
    body: JSON.stringify(payload)
  }).then((response) => response.json());
};

export const patchJSON = (path, payload) => {
  return fetch(`${API_HOST}${path}`, {
    method: "PATCH",
    headers: {
      "content-type": "application/json"
    },
    body: JSON.stringify(payload)
  }).then((response) => response.json());
};
