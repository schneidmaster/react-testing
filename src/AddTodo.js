import React, { useState } from "react";
import { postJSON } from "./api";
import "./AddTodo.css";

const AddTodo = ({ users, addTodo }) => {
  const [adding, setAdding] = useState(false);
  const [title, setTitle] = useState("");
  const [userId, setUserId] = useState("");

  const add = async (e) => {
    e.preventDefault();

    const payload = {
      title,
      userId: parseInt(userId, 10),
      status: "incomplete"
    };

    const newTodo = await postJSON("/todos", payload);
    addTodo(newTodo);
    clear();
  };

  const clear = () => {
    setAdding(false);
    setTitle("");
    setUserId("");
  };

  if (adding) {
    return (
      <form className="add-todo-form" onSubmit={add}>
        <fieldset>
          <legend>Add todo</legend>
          <p>
            <label htmlFor="add-todo-title">Title</label>
            <br />
            <input
              id="add-todo-title"
              onChange={(e) => setTitle(e.target.value)}
              value={title}
            />
          </p>

          <p>
            <label htmlFor="add-todo-user">User</label>
            <br />
            <select
              id="add-todo-user"
              onChange={(e) => setUserId(e.target.value)}
              value={userId}
            >
              <option value="">None</option>
              {users.map((user) => (
                <option key={user.id} value={user.id}>
                  {user.name}
                </option>
              ))}
            </select>
          </p>

          <button type="submit" className="add-todo-btn add-todo-btn-primary">
            Add
          </button>

          <button type="reset" className="add-todo-btn" onClick={clear}>
            Cancel
          </button>
        </fieldset>
      </form>
    );
  } else {
    return (
      <button
        className="add-todo-btn add-todo-btn-primary"
        onClick={() => setAdding(true)}
      >
        Add todo
      </button>
    );
  }
};

export default AddTodo;
