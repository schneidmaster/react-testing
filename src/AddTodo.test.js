import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import AddTodo from "./AddTodo";

// Mock the API so we don't make any real network calls;
jest.mock("./api");
import { postJSON } from "./api";

describe("AddTodo", () => {
  beforeEach(() => {
    // Clear out our mock between each test.
    postJSON.mockClear();
  });

  // Snapshot tests.
  it("renders correctly in button state", () => {
    const users = [
      {
        id: 1,
        name: "Jack User"
      },
      {
        id: 2,
        name: "joseph User"
      }
    ];
    const { container } = render(<AddTodo users={users} />);
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when full form is open", () => {
    const users = [
      {
        id: 1,
        name: "Jack User"
      },
      {
        id: 2,
        name: "joseph User"
      }
    ];
    const { container, getByText } = render(<AddTodo users={users} />);
    const button = getByText("Add todo");
    fireEvent.click(button);
    expect(container).toMatchSnapshot();
  });

  // Test that form fills out and submits successfully.
  // RTL tests like this can straddle the line between
  // unit and integration tests -- they are reliable and
  // fast like unit tests, but the component is actually
  // being exercised in a headless browser so it provides
  // the same security as an integration test.
  it("correctly fills out and submits form", async () => {
    const users = [
      {
        id: 1,
        name: "Jack User"
      },
      {
        id: 2,
        name: "joseph User"
      }
    ];
    const addTodo = jest.fn();
    const { container, getByLabelText, getByText } = render(
      <AddTodo users={users} addTodo={addTodo} />
    );

    // Mock postJSON for creating new todo.
    const newTodo = {
      id: 1,
      title: "Ship the feature ðŸš€",
      status: "incomplete",
      userId: 1
    };
    postJSON.mockImplementation(() => {
      return new Promise((resolve, reject) => {
        resolve(newTodo);
      });
    });

    const startBtn = getByText("Add todo");
    fireEvent.click(startBtn);

    const titleField = getByLabelText("Title");
    fireEvent.change(titleField, { target: { value: "Ship the feature ðŸš€" } });

    const userField = getByLabelText("User");
    fireEvent.change(userField, { target: { value: 1 } });

    const submitBtn = getByText("Add");
    fireEvent.click(submitBtn);

    expect(postJSON).toBeCalledWith("/todos", {
      title: "Ship the feature ðŸš€",
      userId: 1,
      status: "incomplete"
    });

    await wait(() => expect(addTodo).toBeCalledWith(newTodo));
  });
});
