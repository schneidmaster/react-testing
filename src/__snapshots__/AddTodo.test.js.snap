// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo renders correctly in button state 1`] = `
<div>
  <button
    class="add-todo-btn add-todo-btn-primary"
  >
    Add todo
  </button>
</div>
`;

exports[`AddTodo renders correctly when full form is open 1`] = `
<div>
  <form
    class="add-todo-form"
  >
    <fieldset>
      <legend>
        Add todo
      </legend>
      <p>
        <label
          for="add-todo-title"
        >
          Title
        </label>
        <br />
        <input
          id="add-todo-title"
          value=""
        />
      </p>
      <p>
        <label
          for="add-todo-user"
        >
          User
        </label>
        <br />
        <select
          id="add-todo-user"
        >
          <option
            value=""
          >
            None
          </option>
          <option
            value="1"
          >
            Jack User
          </option>
          <option
            value="2"
          >
            joseph User
          </option>
        </select>
      </p>
      <button
        class="add-todo-btn add-todo-btn-primary"
        type="submit"
      >
        Add
      </button>
      <button
        class="add-todo-btn"
        type="reset"
      >
        Cancel
      </button>
    </fieldset>
  </form>
</div>
`;
